### PR TITLE
[2.x] Split driver and connection and types support

### PR DESCRIPTION
**Pull request by CauanCabral**
I rewrite my implementation of `typeConversion` branch.

That's include:

1. Split of Connection and Driver code
2. Support to basic type conversion between elasticsearch type and PHP types
3. Tests

### CODE REVIEW

**Comment by CauanCabral** [comment]
Ok, I can't realize whats breaking tests on Travis. I'll try again later.

In my local env, all they pass:

```PHPUnit 6.5.14 by Sebastian Bergmann and contributors.

...............................................................  63 / 198 ( 31%)
............................................................... 126 / 198 ( 63%)
............................................................... 189 / 198 ( 95%)
.........                                                       198 / 198 (100%)

Time: 11.17 seconds, Memory: 14.00MB

OK (198 tests, 701 assertions)
```

**Comment by CauanCabral** [comment]
I forgot to update `init.php` with `Driver` settings.

Tests are ok now.

Should I close the #235 PR as this one seems a better approach?

File: src/Database/Driver/Elasticsearch.php [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+     * @param array $config config options
+     * @param callable $callback Callback function which can be used to be notified
+     * about errors (for example connection down)
+     */
+    public function __construct(array $config = [], $callback = null)
+    {
+        $config += $this->_baseConfig;
+        $this->_config = $config;
+
+        $esLogger = $config['esLogger'];
```
Comment (markstory) [inline src/Database/Driver/Elasticsearch.php:63]: Any reason to not use `logger`?

File: src/Datasource/Connection.php [hunk 2]
```diff
@@ -88,69 +108,117 @@ public function __construct(array $config = [], $callback = null)
...
-        return true;
+        if (is_string($driver)) {
+            $className = App::className($driver, 'Database/Driver');
+            if (!$className || !class_exists($className)) {
+                throw new MissingDriverException(['driver' => $driver]);
+            }
+
+            $config['esLogger'] = $this->getEsLogger();
+            $callback = $config['callback'] ?? null;
+            $driver = new $className($config, $callback);
```
Comment (markstory) [inline src/Datasource/Connection.php:156]: What would happen if this classname wasn't elasticsearch based? Perhaps this should be constrained to subclasses of the elasticsearch driver class for now?

File: tests/TestCase/ResultSetTest.php [hunk 3]
```diff
@@ -92,13 +99,78 @@ public function testCurrent($resultSets)
...
+        $query->expects($this->once())->method('getRepository')
+            ->will($this->returnValue($index));
+        $index->expects($this->once())
+            ->method('getEntityClass')
+            ->will($this->returnValue(__NAMESPACE__ . '\MyTestDocument'));
+        $index->method('embedded')
+            ->will($this->returnValue([]));
+        $index->expects($this->once())
+            ->method('getConnection')
+            ->will($this->returnValue($connection));
```
Comment (markstory) [inline tests/TestCase/ResultSetTest.php:128]: Could we do a test with fewer mocks? Perhaps an integration test that fetches a document with a variety of types and checks the casted results in the document?

With this many mock is hard to be confident that we're testing actual code here and not just our mocks.

File: src/Database/Driver/Elasticsearch.php (hunk 1 above)
Comment (CauanCabral) [reply src/Database/Driver/Elasticsearch.php:63]: As current `Connection` have both `logger()` and `getEsLogger()` (and both properties), I tried maintain it.

I really don't know if we need both loggers to be configurable, as the `esLogger` is just a wrapper around `logger`.

File: src/Datasource/Connection.php (hunk 2 above)
Comment (CauanCabral) [reply src/Datasource/Connection.php:156]: This is fair, I'll put some validation there.

File: tests/TestCase/ResultSetTest.php (hunk 3 above)
Comment (CauanCabral) [reply tests/TestCase/ResultSetTest.php:128]: Totally agree.

File: src/Database/Driver/Elasticsearch.php (hunk 1 above)
Comment (markstory) [reply src/Database/Driver/Elasticsearch.php:63]: Sorry, I had forgotten about the two loggers. We can clean that up separately.

**Pull request closed without merging**